net/socket.h

```diff
@@ -21,80 +21,215 @@ limitations under the License.
+        }
+
+        bool is_link_local() const {
+            if (is_ipv4()) {
+                return (to_nl() & htonl(0xffff0000)) == htonl(0xa9fe0000);
+            } else {
+                return (addr._in_addr_field[0] & htonl(0xffc00000)) == htonl(0xfe800000);
+            }
+        }
+
+        bool operator==(const IPAddr& rhs) const {
+            return memcmp(this, &rhs, sizeof(rhs)) == 0;
+        }
+
+        bool operator!=(const IPAddr& rhs) const {
+            return !(*this == rhs);
+        }
+
+        const std::string to_string() const {
+            std::string str;
+            char text[INET6_ADDRSTRLEN];
+            if (is_ipv4()) {
+                in_addr ip4;
+                ip4.s_addr = to_nl();
+                inet_ntop(AF_INET, &ip4, text, INET_ADDRSTRLEN);
+            } else {
+                inet_ntop(AF_INET6, &addr, text, INET6_ADDRSTRLEN);
+            }
+            str.assign(text, strlen(text));
+            return str;
+        }
+
+    public:
+        static const IPAddr V6None() {
+            return IPAddr(htonl(0xffffffff), htonl(0xffffffff), htonl(0xffffffff), htonl(0xffffffff));
+        }
+
+        static const IPAddr V6Any() { return IPAddr(in6addr_any); }
+
+        static const IPAddr V6Loopback() { return IPAddr(in6addr_loopback); }
+
+        static const IPAddr V4Broadcast() { return IPAddr(htonl(INADDR_BROADCAST)); }
+
+        static const IPAddr V4Any() { return IPAddr(htonl(INADDR_ANY)); }
+
+        static const IPAddr V4Loopback() { return IPAddr(htonl(INADDR_LOOPBACK)); }
+
+    private:
+        void map_v4(in_addr addr_) {
+            map_v4(addr_.s_addr);
+        }
+
+        void map_v4(uint32_t nl) {
+            addr._in_addr_field[0] = 0x00000000;
+            addr._in_addr_field[1] = 0x00000000;
+            addr._in_addr_field[2] = htonl(0xffff);
+            addr._in_addr_field[3] = nl;
         }
     };
 
-    struct EndPoint
-    {
+    struct EndPoint {
         IPAddr addr;
         uint16_t port = 0;
         EndPoint() = default;
         EndPoint(IPAddr ip, uint16_t port): addr(ip), port(port) {}
-        EndPoint(const struct sockaddr_in& addr_in)
-        {
-            from(addr_in);
-        }
-        sockaddr_in to_sockaddr_in() const
-        {
-            struct sockaddr_in addr_in;
-            addr_in.sin_family = AF_INET;
-            addr_in.sin_addr.s_addr = addr.to_nl();
-            addr_in.sin_port = htons(port);
-            return addr_in;
-        }
-        void from_sockaddr_in(const struct sockaddr_in& addr_in)
-        {
-            addr.from_nl(addr_in.sin_addr.s_addr);
-            port = ntohs(addr_in.sin_port);
-        }
-        void from(const struct sockaddr_in& addr_in)
-        {
-            from_sockaddr_in(addr_in);
+        explicit EndPoint(const sockaddr_storage* sock_addr) {
+            from_sockaddr_storage(sock_addr);
+        }
+        bool is_ipv4() const {
+            return addr.is_ipv4();
+        };
```

Comment:
`sockaddr_storage` is defined in `#include <netinet/in.h>`, and is wide enough to hold any type of socket address